tools/publish-aos.js

```diff
@@ -1,54 +1,170 @@
+      const txId = idWithExtension.split('.').shift(); // Remove the file extension (e.g., .md)
+
+      return txId;
+    })
+    .filter((txId) => txId.length === 43); // Filter for valid IDs with length 43
+
+  return txIds;
+}
 async function main() {
+  // get the last published source code from the changelogs folder (can check the date on it, check for multiple dates) and
+  // fetch the code from arweave, then compare it with the current code
+
+  const publishedSourceCodeIds = getPublishedSourceCodeIds();
+  const sourceCodeTransactions = await arweave.api
+    .post('/graphql', {
+      query: `{transactions(
+        ids:[ ${publishedSourceCodeIds.map((id) => `"${id}"`).join(',')}]) {
+			edges {
+				node {
+					id
+					tags {
+						name
+						value
+					}
+          block {
+            timestamp
+          }
+				}
+			}
+		}
+        }`,
+    })
+
+    .then((res) => {
+      return res.data.data.transactions.edges
+        .map((edge) => edge.node)
+        .sort((a, b) => b.block.timestamp - a.block.timestamp);
+    });
+
+  const lastPublishedSourceCodeId = sourceCodeTransactions[0].tags.find(
+    (tag) => tag.name === 'Original-Tx-Id',
+  )?.value;
+
+  const lastPublishedSourceCode = await arweave.transactions
+    .getData(lastPublishedSourceCodeId, { decode: true, string: true })
+    .catch((error) => {
+      console.error(
+        'Error fetching last published source code, unable to compare with current code. \n',
+        error,
+      );
+    });
+
   const bundledLua = fs.readFileSync(
     path.join(__dirname, '../dist/aos-bundled.lua'),
     'utf-8',
   );
-  const wallet = fs.readFileSync(path.join(__dirname, 'key.json'), 'utf-8');
+
+  if (lastPublishedSourceCode === bundledLua) {
+    console.log('No changes in source code detected, skipping publishing');
+    process.exit(0); // Exit with code 0 to indicate success
+  }
+  let wallet = process.env.ARWEAVE_PUBLISHING_KEY;
+  if (!wallet) {
+    try {
+      wallet = fs.readFileSync(path.join(__dirname, 'key.json'), 'utf-8');
+    } catch (error) {
+      console.error(
+        'No key found, please set ARWEAVE_PUBLISHING_KEY environment variable or provide a key.json file',
+      );
+      return;
+    }
+  }
   const jwk = JSON.parse(wallet);
   const address = await arweave.wallets.jwkToAddress(jwk);
+  const signer = new ArweaveSigner(jwk);
+  const turbo = TurboFactory.authenticated({ signer });
 
-  const tx1 = await arweave.createTransaction({ data: bundledLua }, jwk);
-  tx1.addTag('App-Name', 'aos-LUA');
-  tx1.addTag('App-Version', '0.0.1');
-  tx1.addTag('Content-Type', 'text/x-lua');
-  tx1.addTag('Author', 'Permanent Data Solutions');
-  await arweave.transactions.sign(tx1, jwk);
-  await arweave.transactions.post(tx1);
+  const publishingTags = Object.entries({
+    'App-Name': 'aos-LUA',
+    'App-Version': '0.0.1',
+    'Content-Type': 'text/x-lua',
+    Author: 'Permanent Data Solutions',
+    Changelog: changelog,
```

Comment:
i'm reluctant to include the changelog in the tags - instead can we include the `Git-Hash`?